Add container apps job specs

anandanthony opened the pull request:
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [x] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

anandanthony committed:
Adding Container App Jobs specs for review

anandanthony committed:
Adding Container App Jobs specs for review

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

jijohn14 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/examples/ContainerAppsJob_CreateorUpdate.json:
```diff
@@ -0,0 +1,143 @@
+      "location": "East US",
+      "properties": {
+        "environmentId": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/managedEnvironments/demokube",
+        "workloadProfileType": "GeneralPurpose",
```
remove since this is a premium related property

jijohn14 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/examples/ContainerAppsJob_Run.json:
```diff
@@ -0,0 +1,11 @@
+    "api-version": "2022-11-01-preview"
+  },
+  "responses": {
+    "202": {}
```
Is there no operation id to get status of job run

anandanthony replied:
@jijohn14 currently there is no operation Id associated with a single run. We had discussions around it.
Adding @vinisoto @trajkobal

jijohn14 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/examples/ContainerAppsJob_Terminate.json:
```diff
@@ -0,0 +1,11 @@
+    "api-version": "2022-11-01-preview"
+  },
+  "responses": {
+    "202": {}
```
Same. How can user confirm the operation is completed

trajkobal replied:
With the job name in the path parameter with a status call

jijohn14 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,788 @@
+          ]
+        },
+        "triggerType": {
+          "description": "Type of the job - Schedule | Event | Manual",
```
Remove values from description

mentat9 replied:
Note you can add descriptions for each enum value using x-ms-enum. This can be very helpful to clients in some cases.

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,788 @@
+            "description": "Name of the Container Apps Job.",
+            "required": true,
+            "type": "string",
+            "pattern": "^.*$"
```
@jijohn14, updated, had just added this as it is mandatory to have pattern, but we are going to validate the name at control plane. Updated the pattern now to be more specific on what is expected.

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony commented:
@leni-msft requesting you to please review this PR. Also, we can see that the SDK checks fail which look to be an issue with how the changes are being built rather than the files we have added.

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

anandanthony committed:
Adding Container App Jobs specs for review | fix validations

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+        },
+        "identity": {
+          "description": "Managed identities needed by a container app job to interact with other Azure services to not maintain any secrets or credentials in code.",
+          "$ref": "../../../../../common-types/resource-management/v3/managedidentity.json#/definitions/ManagedServiceIdentity"
```
>v3 [](http://example.com/codeflow?start=67&length=2)

Recommend using v5 identity types.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+      "properties": {
+        "identity": {
+          "description": "Managed identities needed by a container app job to interact with other Azure services to not maintain any secrets or credentials in code.",
+          "$ref": "../../../../../common-types/resource-management/v3/managedidentity.json#/definitions/ManagedServiceIdentity"
```
>v3 [](http://example.com/codeflow?start=67&length=2)

Recommend v5

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+        "operationId": "ContainerAppsJobs_ListBySubscription",
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
```
>v3 [](http://example.com/codeflow?start=69&length=2)

Recommend using current v4 common types. Comment applies throughout.

anandanthony replied:
@mentat9 we updated the references to /v4 by adding the types.json file to this PR, but can see that the Swagger validation fails. How we can address this?

mentat9 replied:
This may have been due to a bad change in v4 that has since been reverted. See if it goes away with a refresh from main (you can diff v3 and v4 and see that they are now the same). Either way is fine: I won't block on it.

anandanthony replied:
@mentat9 we created a different PR today to add the new/updated version to the "release-app-Microsoft.App-2022-11-01-preview" branch and got the same Breaking change validation error. We added the files in the "/specification/common-types/resource-management" from the main branch.
https://github.com/Azure/azure-rest-api-specs/pull/22114 

Also, wanted to know why the SDK validations were skipped for the given PR?

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+              "$ref": "#/definitions/ContainerAppsJob"
+            }
+          },
+          "404": {
```
>404 [](http://example.com/codeflow?start=11&length=3)

Remove this and use default branch for all errors.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+          }
+        ],
+        "responses": {
+          "201": {
```
>201 [](http://example.com/codeflow?start=11&length=3)

This response needs to include the resource payload (with provisioningState).

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+        "responses": {
+          "201": {
+            "description": "Container Apps Job create has been started.",
+            "headers": {
```
>headers [](http://example.com/codeflow?start=13&length=7)

Don't need to declare headers. Comment applies throughout.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          }
+        ],
+        "responses": {
```
>responses [](http://example.com/codeflow?start=9&length=9)

Need 200 (and optionally 204) for final response.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+              }
+            }
+          },
+          "404": {
```
>404 [](http://example.com/codeflow?start=11&length=3)

Use default branch for all errors.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+        "properties": {
+          "type": "object",
+          "properties": {
+            "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=13&length=17)

provisioningState shouldn't be PATCHable.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+                "modelAsString": true
+              }
+            },
+            "environmentId": {
```
>environmentId [](http://example.com/codeflow?start=13&length=13)

Mutability should include update if it's part of a PATCH payload.

anandanthony replied:
added "update" as mutability.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+              "$ref": "./CommonDefinitions.json#/definitions/ContainerAppsJobTemplate",
+              "description": "Container Apps job definition."
+            },
+            "outboundIpAddresses": {
```
>outboundIpAddresses [](http://example.com/codeflow?start=13&length=19)

PATCHable properties can't be readOnly.

mentat9 replied:
Turns out it's actually allowed so payload definition can be shared across PUT and PATCH, but in this case where it's just PATCH, it doesn't make sense.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+              },
+              "readOnly": true
+            },
+            "eventStreamEndpoint": {
```
>eventStreamEndpoint [](http://example.com/codeflow?start=13&length=19)

Same

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          }
+        ],
+        "responses": {
```
Need 200 for final response.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+            "pattern": "^[-\\w\\._\\(\\)]+$"
+          }
+        ],
+        "responses": {
```
>responses [](http://example.com/codeflow?start=9&length=9)

Need 200 for final response.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/containerAppsJobs/{containerAppsJobName}/run": {
```
>run [](http://example.com/codeflow?start=137&length=3)

I think start and cancel are better action names for this and the next POST action. Just offering an opinion, this wouldn't block ARM signoff.

ruslany replied:
Agree. Should we call this start and rename "terminate" to "stop"

anandanthony replied:
@vinisoto please suggest. Currently we have run and terminate as the actions.

vinisoto replied:
I'm ok with the change.

anandanthony replied:
@mentat9 thanks for the suggestion. Updated this.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+        "triggerType"
+      ],
+      "properties": {
+        "secrets": {
```
>secrets [](http://example.com/codeflow?start=9&length=7)

Your backend needs to null or empty out this property in all GET, PUT, and PATCH payloads.

anandanthony replied:
@mentat9 we will handle this in the backend logic.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+      "description": "Container Apps Job Dapr configuration.",
+      "type": "object",
+      "properties": {
+        "enabled": {
```
>enabled [](http://example.com/codeflow?start=9&length=7)

Recommend enums (in general) over booleans, e.g. activeState: { Active, Disabled } or similar, or even enabled: { True, False } if you prefer. Ref: [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review) for more details.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+          "description": "Dapr application identifier",
+          "type": "string"
+        },
+        "appProtocol": {
```
>appProtocol [](http://example.com/codeflow?start=9&length=11)

You can provide descriptions for each enum value using x-ms-enum extension.

mentat9 commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,861 @@
+            "modelAsString": true
+          }
+        },
+        "enableApiLogging": {
```
>enableApiLogging [](http://example.com/codeflow?start=9&length=16)

Consider using enum instead of boolean.

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anthonychu commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerAppsJobs.json:
```diff
@@ -0,0 +1,781 @@
+      }
+    },
+    "ReplicaCount": {
+      "description": "Minimum number of successful replica completions before overall job completion.",
```
This description is different than spec. In spec this is the number of replicas in total to run in a job execution.

anandanthony replied:
@anthonychu we discussed on this and updated this as per the following suggestion: https://github.com/Azure/azure-rest-api-specs/pull/21846/files/feaac967641b198543cd57b4fb2644b3b3a80a18#diff-739c3ca30297197a20a321984788ec7f5189f3f637da962aae461dc834f5613c

@trajkobal do we need to revert it back?

anthonychu replied:
If this maps to `completions` in k8s, then I think `replicaCount` is not the right name for it. I suggest we consider `completions` or `replicaCompletionCount`.

trajkobal replied:
I agree with you @anthonychu. Now we have to convince @vinisoto and @anandanthony

trajkobal replied:
> @anthonychu we discussed on this and updated this as per the following suggestion: https://github.com/Azure/azure-rest-api-specs/pull/21846/files/feaac967641b198543cd57b4fb2644b3b3a80a18#r1048686943
> 
> @trajkobal do we need to revert the description back?

No, unless we have a better way to explain it. The spec description seemed a bit obscure to me.

anandanthony replied:
@anthonychu we discussed on the description given here. Basically, the property "replicaCount" would be used to define the minimum number of replicas that a job would get created for a job and completed on a job's execution. In case anyone of the replicas fail, another replica would get created and executed. This would happen till all the replicas (as per the replicaCount) have completed their execution (before replicaTimeout). Hence, we went ahead with this description and name.

anandanthony replied:
@anthonychu updated this property as per our discussion.

anthonychu commented on specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/examples/ContainerAppsJob_CreateorUpdate.json:
```diff
@@ -0,0 +1,207 @@
+          "replicaTimeout": 10,
+          "replicaRetryLimit": 10,
+          "manualTriggerConfig": {
+            "replicaCount": 1,
```
`replicaCount` should be `>= parallelism`

anandanthony replied:
@anthonychu this was modified as per the following suggestion - https://github.com/Azure/azure-rest-api-specs/pull/21846/files/feaac967641b198543cd57b4fb2644b3b3a80a18#r1048685332

@trajkobal requesting your input on this.

trajkobal replied:
It can go both ways. 

- If the replica count (completions) is less than parallelism (e.g. you execute 2 parallel replicas and only need one of them to complete, then the job is done when one replica finishes. This is more like providing redundancy.
- If the replica count is more than parallelism then it will repeat replica count / parallelism number of times. i.e run 10 replicas 2 at a time way

anthonychu reviewed (commented):
@anandanthony A couple of small comments.

anandanthony committed:
Adding Container App Jobs specs for review | address comments

anandanthony committed:
updating sub-resource name to jobs from containerAppsJobs

anandanthony committed:
updating sub-resource name to jobs from containerAppsJobs

anandanthony committed:
updating sub-resource name to jobs from containerAppsJobs

anandanthony committed:
updating sub-resource name to jobs from containerAppsJobs

anandanthony committed:
updating sub-resource name to jobs from containerAppsJobs

leni-msft commented:
@jianyexi to help check breaking change exceptions. @msyyc @tadelesh to help review SDK generation failure.

tadelesh commented:
> @jianyexi to help check breaking change exceptions. @msyyc @tadelesh to help review SDK generation failure.

Duplicate schema of `Resource` causes SDK generation failure.

anandanthony commented:
@tadelesh thank you for your comment, can you please confirm if the version difference in the following files is causing the SDK generation failure?
[Microsoft.App/preview/2022-11-01-preview/Jobs.json](https://github.com/anandMicro/azure-rest-api-specs/blob/0f37f1671a44ec4bd609539dcf5971a9cc1c6f50/specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/Jobs.json#L581)
[Microsoft.App/preview/2022-11-01-preview/ContainerApps.json](https://github.com/anandMicro/azure-rest-api-specs/blob/0f37f1671a44ec4bd609539dcf5971a9cc1c6f50/specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerApps.json#L527)
We updated the `Resource` version due to a previous comment. In case updating the version is not mandatory, we can revert back the version changes.

tadelesh commented:
> @tadelesh thank you for your comment, can you please confirm if the version difference in the following files is causing the SDK generation failure? [Microsoft.App/preview/2022-11-01-preview/Jobs.json](https://github.com/anandMicro/azure-rest-api-specs/blob/0f37f1671a44ec4bd609539dcf5971a9cc1c6f50/specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/Jobs.json#L581) [Microsoft.App/preview/2022-11-01-preview/ContainerApps.json](https://github.com/anandMicro/azure-rest-api-specs/blob/0f37f1671a44ec4bd609539dcf5971a9cc1c6f50/specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerApps.json#L527) We updated the `Resource` version due to a previous comment. In case updating the version is not mandatory, we can revert back the version changes.

I'm not sure it is the only discrepancy, but you could revert it or consolidate them to the new version and see if the SDK generation will pass.

anandanthony committed:
revert resource management version

anandanthony committed:
revert resource management version

anandanthony committed:
revert resource management version

anandanthony commented:
> > @tadelesh thank you for your comment, can you please confirm if the version difference in the following files is causing the SDK generation failure? [Microsoft.App/preview/2022-11-01-preview/Jobs.json](https://github.com/anandMicro/azure-rest-api-specs/blob/0f37f1671a44ec4bd609539dcf5971a9cc1c6f50/specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/Jobs.json#L581) [Microsoft.App/preview/2022-11-01-preview/ContainerApps.json](https://github.com/anandMicro/azure-rest-api-specs/blob/0f37f1671a44ec4bd609539dcf5971a9cc1c6f50/specification/app/resource-manager/Microsoft.App/preview/2022-11-01-preview/ContainerApps.json#L527) We updated the `Resource` version due to a previous comment. In case updating the version is not mandatory, we can revert back the version changes.
> 
> I'm not sure it is the only discrepancy, but you could revert it or consolidate them to the new version and see if the SDK generation will pass.

@tadelesh we have reverted the version changes and now the breaking change validation is successful.
